# Amplicon placement

## Install UShER 

### Conda Local Build
Follow these steps to setup local conda environment
```
git clone https://github.com/yatisht/usher.git
cd usher/install
conda env create -f environment.yml
conda activate usher
cd ..
mkdir build
cd build
git clone https://github.com/oneapi-src/oneTBB
cd oneTBB
git checkout cc2c04e2f5363fb8b34c10718ce406814810d1e6
cd ..
cmake  -DTBB_DIR=${PWD}/oneTBB  -DCMAKE_PREFIX_PATH=${PWD}/oneTBB/cmake -DCMAKE_BUILD_TYPE=Release ..
make -j
cd ..

```

followed by, if on a MacOS system:
```
rsync -aP rsync://hgdownload.soe.ucsc.edu/genome/admin/exe/macOSX.x86_64/faToVcf .
chmod +x faToVcf
mv faToVcf build/
```

or if on a Linux system:
```
rsync -aP rsync://hgdownload.soe.ucsc.edu/genome/admin/exe/linux.x86_64/faToVcf .
chmod +x faToVcf
mv faToVcf build/
```

Executables will be located in the build directory. Make sure theyâ€™re on your path for your system as appropriate, or that you modify your commands to indicate their location.
```
export PATH=$PATH:/path/to/install/usher/build/
```

### Installation scripts
```
git clone https://github.com/yatisht/usher.git
cd usher
```

For Ubuntu 18.04 and above (requires sudo privileges):
```
./install/installUbuntu.sh
export PATH=$PATH:/path/to/install/usher/build/
```

Then change into `scripts/amplicon` directory and run the following installation script to install amplicon placement specific dependencies.
```
cd scripts/amplicon
./amplicon_installUbuntu.sh
export PATH=$PATH:/path/to/install/usher/scripts/amplicon/build/
```

## Generate amplicon variant call file
Change into the `scripts/amplicon` directory for amplicon placement workflow.
```
cd scripts/amplicon
```

- Download SARS-CoV2 reference
```
wget https://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/bigZips/wuhCor1.fa.gz
gunzip wuhCor1.fa.gz
```

- Download sites to mask 
```
wget https://raw.githubusercontent.com/W-L/ProblematicSites_SARS-CoV2/master/problematic_sites_sarsCov2.vcf
```
- Download latest public tree
```
wget http://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/UShER_SARS-CoV-2/public-latest.all.masked.pb.gz
gunzip public-latest.all.masked.pb.gz
```

- Make sure your amplicon reads file (FASTQ) is in current directory (`scripts/amplicon`)

- Run `get_variants.py` script below to generate VCF for amplicon reads (`data/amplicons.vcf`) and an MSA FASTA file (`data/amplicons.fa`) which is used to obtain amplicon read alignment coordinate positions.
```
python3 get_variants.py -f <amplicon_reads.fastq> [options]
```
Note: `-f` is a required flag to give amplicon reads FASTQ file

Options:
- `-h`|`--help`: Get usage information
- `-a`|`--align_with`: Specify to use alternate aligner. Default alignment with `minimap2` if no flag is given.
- `-r`|`--remove`: Remove duplicate amplicon reads in given FASTQ file 

Once this script is run successfully, `data/amplicons.fa` and `data/amplicons.vcf` will be generated and needed for the next placement command.


## Placing amplicon sequences using UShER
<br>
Now use UShER to place the amplicon sequences on the tree previously downloaded, using the following command:

```
amplicon -i <tree.pb> -f data/amplicons.fa -v data/amplicons.vcf -d output
```

Required Options:
- `-i`: MAT protobuf to place amplicon samples onto
- `-v`: the generated vcf file from the previous steps
- `-f`: MSA FASTA file generated by the previous step that contains amplicon coordinate information necessary for placement
- `-d`: (Optional) output directory, defaults to current directory (if flag not given, files will be output to current directory)         

### Output files
The following two output files will be generated after placement and written to the output directory or current directory if `-d` flag not given.
- `final-tree.nh`:  A generated Newick file of the tree 
- `clade.txt`: Information about likely clade assignments for each amplicon sample, relative to the number of equally parsimonious placements possible for the given sample. 
 
